src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -1148,21 +1147,33 @@ internal SingleLookupResult CheckViability(Symbol symbol, int arity, LookupOptio
                 return LookupResult.Inaccessible(symbol, diagInfo);
             }
             else if (!InCref &&
-                     !this.IsAccessible(unwrappedSymbol,
+                    !this.IsAccessible(unwrappedSymbol,
                                         RefineAccessThroughType(options, accessThroughType),
                                         out inaccessibleViaQualifier,
                                         ref useSiteDiagnostics,
                                         basesBeingResolved))
             {
-                if (inaccessibleViaQualifier)
-                {
-                    diagInfo = diagnose ? new CSDiagnosticInfo(ErrorCode.ERR_BadProtectedAccess, unwrappedSymbol, accessThroughType, this.ContainingType) : null;
-                }
-                else
-                {
-                    var unwrappedSymbols = ImmutableArray.Create<Symbol>(unwrappedSymbol);
-                    diagInfo = diagnose ? new CSDiagnosticInfo(ErrorCode.ERR_BadAccess, new[] { unwrappedSymbol }, unwrappedSymbols, additionalLocations: ImmutableArray<Location>.Empty) : null;
-                }
+                 bool friendRefNotEqualToThis = unwrappedSymbol != null && unwrappedSymbol.ContainingAssembly.GetInternalsVisibleToPublicKeys(this.ContainingType.ContainingAssembly.Name).Any()
+                            && unwrappedSymbol.DeclaredAccessibility.Equals(Accessibility.Internal);
+
+                 foreach (ImmutableArray<byte> key in unwrappedSymbol.ContainingAssembly.GetInternalsVisibleToPublicKeys(this.ContainingType.ContainingAssembly.Name))
+                 {
+                     friendRefNotEqualToThis = key.SequenceEqual(this.ContainingType.ContainingAssembly.Identity.PublicKey) ? false : friendRefNotEqualToThis;
+                 }
```

Comment:
Can be combined with `friendRefNotEqualToThis` assignment above:
```
bool friendRefNotEqualToThis = unwrappedSymbol != null && ... &&
        unwrappedSymbol.ContainingAssembly.GetInternalsVisibleToPublicKeys(
            this.ContainingAssembly.Name).Any((key, publicKey) =>
                key.SequenceEqual(publicKey),
                this.ContainingAssembly.Identity.PublicKey);

Reply:
The body of the lambda depends on the lambda parameters `key` and `publicKey` only.